syft/pkg/cataloger/nix/parse_nix_store_path_test.go

```diff
@@ -0,0 +1,284 @@
+package nix
+
+import (
+	"path"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func Test_findVersionIsh(t *testing.T) {
+	tests := []struct {
+		name    string
+		input   string
+		wantIdx int
+		wantEx  string
+		wantPr  string
+	}{
+		{
+			name:    "no version",
+			input:   "5q7vxm9lc4b9hifc3br4sr8dy7f2h0qa-source",
+			wantIdx: -1,
+			wantEx:  "",
+			wantPr:  "",
+		},
+		{
+			name:    "semver with overbite into output",
+			input:   "/nix/store/h0cnbmfcn93xm5dg2x27ixhag1cwndga-glibc-2.34-210-bin",
+			wantIdx: 50,
+			wantEx:  "2.34-210-bin",
+			wantPr:  "210-bin",
+		},
+		{
+			name:    "multiple versions",
+			input:   "5zzrvdmlkc5rh3k5862krd3wfb3pqhyf-perl5.34.1-TimeDate-2.33",
+			wantIdx: 53,
+			wantEx:  "2.33",
+			wantPr:  "",
+		},
+		{
+			name:    "name ends with number",
+			input:   "55nswyz8335lk954y1ccx6as2jbq1z8f-libfido2-1.10.0",
+			wantIdx: 42,
+			wantEx:  "1.10.0",
+			wantPr:  "",
+		},
+		{
+			name:    "major-minor only",
+			input:   "q8gnp7r8475p52k9gmdzsrcddw5hirbn-gdbm-1.23",
+			wantIdx: 38,
+			wantEx:  "1.23",
+			wantPr:  "",
+		},
+		{
+			name:    "0-prefixed version field",
+			input:   "r705jm2icczpnmfccby3fzfrckfjakx3-perl5.34.1-URI-5.05",
+			wantIdx: 48,
+			wantEx:  "5.05",
+			wantPr:  "",
+		},
+		{
+			name:    "prerelease with alpha prefix",
+			input:   "v48s6iddb518j9lc1pk3rcn3x8c2ff0j-bash-interactive-5.1-p16",
+			wantIdx: 50,
+			wantEx:  "5.1-p16",
+			wantPr:  "p16",
+		},
+		{
+
+			name:    "0-major version",
+			input:   "x2f9x5q6qrs6cssx09ylxqyg9q2isi1z-aws-c-http-0.6.15",
+			wantIdx: 44,
+			wantEx:  "0.6.15",
+			wantPr:  "",
+		},
+		{
+
+			name: "several version fields",
+			// note: this package version is fictitious
+			input:   "z24qs6f5d1mmwdp73n1jfc3swj4v2c5s-krb5-1.19.3.9.10",
+			wantIdx: 38,
+			wantEx:  "1.19.3.9.10",
+			wantPr:  "",
+		},
+		{
+
+			name:    "skip drv + major only version",
+			input:   "z0fqylhisz47krxv8fd0izm1i2qbswfr-readline63-006.drv",
+			wantIdx: 44,
+			wantEx:  "006",
+			wantPr:  "",
+		},
+		{
+
+			name:    "prerelease with multiple dashes",
+			input:   "zkgyp2vra0bgqm0dv1qi514l5fd0aksx-bash-interactive-5.1-p16-man",
+			wantIdx: 50,
+			wantEx:  "5.1-p16-man",
+			wantPr:  "p16-man",
+		},
+		{
+
+			name:    "date as major version",
+			input:   "0amf0d1dymv9gqcyhhjb9j0l8sn00c56-libedit-20210910-3.1",
+			wantIdx: 41,
+			wantEx:  "20210910-3.1",
+			wantPr:  "3.1",
+		},
+		{
+
+			name:    "long name",
+			input:   "0296qxvn30z9b2ah1g5p97k5wr9k8y78-busybox-static-x86_64-unknown-linux-musl-1.35.0",
+			wantIdx: 74,
+			wantEx:  "1.35.0",
+			wantPr:  "",
+		},
```

Comment:
There's also a unstable-yyyy-mm-dd convention, if nixpkgs pulls in a version that hasn't had a release recently. I feel like it should also be part of the unit tests.

Reply:
Just so I understand, is this the behavior that needs to be updated and tested for?
> If a package is not a release but a commit from a repository, then the version attribute must be the date of that (fetched) commit. The date must be in "unstable-YYYY-MM-DD" format.

https://nixos.org/manual/nixpkgs/stable/#sec-package-naming (I'll add this in a comment in the code for future folks to see)

Reply:
We should include an additional test for packages / paths that use unstable-yyyy-mm-dd.
I didn't try out piping such a path through the regex yesterday, but even if it works right now, i feel like it should be tested separately.

Example: https://github.com/NixOS/nixpkgs/blob/798e23beab9b5cba4d6f05e8b243e1d4535770f3/pkgs/servers/webdav-server-rs/default.nix#L11

In this case I'd expect the "package name" to be `webdav-server-rs`, and the version to be `unstable-2021-08-16`.

Reply:
Updated 🙌  https://github.com/anchore/syft/pull/1696/commits/46bf1fc55a3a70f1c16c3b44c37ceff32edd9655